### PR TITLE
Refactor graphql association resolvers

### LINKED ISSUE #4353: Graphql - querying deeply nested "has Many" One to Many relationship throws an exception
**Describe the bug**
Objects located deeper than 3 levels with "has Many" one-to-many relationship always throw exception when querying using GraphQL.

**Steps to reproduce the behavior**
I have deeply nested objects.  
building -> apartments (has Many relationship)
apartment -> rooms (has Many relationship)
The models are described in another bug (with the node that rooms are many-to-many relationship) here: https://github.com/strapi/strapi/issues/4288

**Request:**

```graphql
query {
 building(id:1) {
      number
 	apartments {
      rooms {
       number
     }
   }      
 }   
}
```

**Result:**

```json
{
  "errors": [
    {
      "message": "Your filters contain a field 'undefined' that doesn't appear on your model definition nor it's relations",
      "locations": [
        {
          "line": 5,
          "column": 7
        }
      ],
      "path": [
        "building",
        "apartments",
        0,
        "rooms"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "status": 400,
          "stacktrace": [
            "Error: Your filters contain a field 'undefined' that doesn't appear on your model definition nor it's relations",
            "    at getAssociationFromFieldKey (/home/dan/strapi/project/node_modules/strapi-utils/lib/buildQuery.js:40:19)",
            "    at /home/dan/strapi/project/node_modules/strapi-utils/lib/buildQuery.js:120:50",
            "    at Array.map (<anonymous>)",
            "    at buildQuery (/home/dan/strapi/project/node_modules/strapi-utils/lib/buildQuery.js:119:8)",
            "    at Object.find (/home/dan/strapi/project/node_modules/strapi-hook-bookshelf/lib/queries.js:76:14)",
            "    at Object.fetchAll (/home/dan/strapi/project/node_modules/strapi-plugin-content-manager/services/ContentManager.js:26:47)",
            "    at Object.makeQuery (/home/dan/strapi/project/node_modules/strapi-plugin-graphql/services/Loaders.js:172:7)",
            "    at /home/dan/strapi/project/node_modules/strapi-plugin-graphql/services/Loaders.js:56:37",
            "    at Array.map (<anonymous>)",
            "    at /home/dan/strapi/project/node_modules/strapi-plugin-graphql/services/Loaders.js:56:19",
            "    at dispatchQueueBatch (/home/dan/strapi/project/node_modules/dataloader/index.js:246:22)",
            "    at dispatchQueue (/home/dan/strapi/project/node_modules/dataloader/index.js:233:5)",
            "    at /home/dan/strapi/project/node_modules/dataloader/index.js:70:20",
            "    at processTicksAndRejections (internal/process/task_queues.js:75:11)"
          ]
        }
      }
    },
    {
      "message": "Your filters contain a field 'undefined' that doesn't appear on your model definition nor it's relations",
      "locations": [
        {
          "line": 5,
          "column": 7
        }
      ],
      "path": [
        "building",
        "apartments",
        1,
        "rooms"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "status": 400,
          "stacktrace": [
            "Error: Your filters contain a field 'undefined' that doesn't appear on your model definition nor it's relations",
            "    at getAssociationFromFieldKey (/home/dan/strapi/project/node_modules/strapi-utils/lib/buildQuery.js:40:19)",
            "    at /home/dan/strapi/project/node_modules/strapi-utils/lib/buildQuery.js:120:50",
            "    at Array.map (<anonymous>)",
            "    at buildQuery (/home/dan/strapi/project/node_modules/strapi-utils/lib/buildQuery.js:119:8)",
            "    at Object.find (/home/dan/strapi/project/node_modules/strapi-hook-bookshelf/lib/queries.js:76:14)",
            "    at Object.fetchAll (/home/dan/strapi/project/node_modules/strapi-plugin-content-manager/services/ContentManager.js:26:47)",
            "    at Object.makeQuery (/home/dan/strapi/project/node_modules/strapi-plugin-graphql/services/Loaders.js:172:7)",
            "    at /home/dan/strapi/project/node_modules/strapi-plugin-graphql/services/Loaders.js:56:37",
            "    at Array.map (<anonymous>)",
            "    at /home/dan/strapi/project/node_modules/strapi-plugin-graphql/services/Loaders.js:56:19",
            "    at dispatchQueueBatch (/home/dan/strapi/project/node_modules/dataloader/index.js:246:22)",
            "    at dispatchQueue (/home/dan/strapi/project/node_modules/dataloader/index.js:233:5)",
            "    at /home/dan/strapi/project/node_modules/dataloader/index.js:70:20",
            "    at processTicksAndRejections (internal/process/task_queues.js:75:11)"
          ]
        }
      }
    }
  ],
  "data": {
    "building": {
      "number": "12",
      "apartments": [
        {
          "rooms": null
        },
        {
          "rooms": null
        }
      ]
    }
  }
}
```

**Expected behavior**
It should return the rooms as well.

 
**System**
Node.js version: v12.12.0
NPM version: 6.11.3
Strapi version: 3.0.0-beta.17.2
Database: sqlite
Operating system: Ubuntu

### LINKED ISSUE #8948: Admin UI filter breaks when trying to filter by relation field (mongodb)
### Describe the bug
I am getting error when trying to filter by a relation field of other collection.

Only happens when I set **mongodb** as my database, works fine with sqlite and mysql. 
And it seems to happen only when there a lot of entries to filter on the collection.

### Steps to reproduce the behavior
1. Create a project based on **mongodb** 
2. Create a collection with some string field
3. Create another collection with a relation field to the first collection
4. Go to that second collection internal page (List page)
5. Try to filter by the relation field defined ([Field] is "something")
6. Error arises `"Something went wrong" => Cannot read property 'id' of null`

### Expected behavior
They filter should've worked.

### Screenshots

Filtering by relation field:
Screen_Shot_2020-12-19_at_03_20_49

Error:
Screen Shot 2020-12-19 at 02 45 22

### PR DESCRIPTION
**Pull request by alexandrebodin**
### What does it do?

Refactor graphql association resolvers

### Why is it needed?

Too many bugs

### Related issue(s)/PR(s)

Fix https://github.com/strapi/strapi/issues/9035
Fix https://github.com/strapi/strapi/issues/5608
Fix #4353
Fix #8948

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
Will Test, ty @alexandrebodin

File: packages/strapi-plugin-graphql/services/__tests__/data-loaders.test.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+      global.strapi = {
+        query() {
+          return { find, findOne };
+        },
+      };
+
+      await dataLoaders.makeQuery(uid, { single: true, filters });
+
+      expect(findOne).toHaveBeenCalledWith(filters, []);
```
Comment (petersg83) [inline packages/strapi-plugin-graphql/services/__tests__/data-loaders.test.js:33]: You mocked the find function but don't use it no ? You may check that it is not called

File: packages/strapi-plugin-graphql/services/data-loaders.js [hunk 2]
```diff
@@ -35,170 +35,37 @@ module.exports = {
...
-          .slice(skip, skip + limit);
-      }
-
-      return data
-        .filter(entry => entry !== undefined)
-        .filter(entry => ids.map(id => id.toString()).includes(entry[ref.primaryKey].toString()))
-        .slice(skip, skip + limit);
-    });
+  serializeKey(key) {
+    return _.isObjectLike(key) ? JSON.stringify(_.cloneDeep(key)) : key;
```
[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/strapi-plugin-graphql/services/data-loaders.js:56]: I didn't change the code put I will clean it up

File: packages/strapi-plugin-graphql/services/utils.js [hunk 3]
```diff
@@ -79,7 +79,7 @@ const amountLimiting = (params = {}) => {
 
   if (!amountLimit) return params;
 
-  if (!params.limit || params.limit === -1 || params.limit > amountLimit) {
+  if (_.isNil(params.limit) || params.limit === -1 || params.limit > amountLimit) {
```
Comment (soupette) [inline packages/strapi-plugin-graphql/services/utils.js:82]: Why don't we simplify the statement `params.limit === -1` to `params.limit < 0` to get rid of the else if condition? 
I believe that we should have the same logic is we query with a limit of `-2` than with `-1`, what do you think?

Comment (alexandrebodin) [reply packages/strapi-plugin-graphql/services/utils.js:82]: Actually  -1 is a specific value to ask for everything. other negative values are invalid

Comment (soupette) [reply packages/strapi-plugin-graphql/services/utils.js:82]: Ok I see! thanks for your response

**Review by petersg83 — APPROVED** [review]
LGTM :)

**Comment by abdonrd** [comment]
@alexandrebodin @derrickmehaffy @petersg83 @soupette As the relationships becomes much slower from Strapi `v3.4.1` to `v3.4.2`... Could it be because of something in this PR?

More info: https://github.com/strapi/strapi/issues/8552#issuecomment-770388720

**Comment by alexandrebodin** [comment]
this PR generally improves performances; but in some scenarios make it a slower as it fixes the querying that sometimes require more processing to be accurate.

**Pull request merged**